modules/kafka/src/main/scala/com/snowplowanalytics/snowplow/sinks/kafka/KafkaSink.scala

```diff
@@ -7,49 +7,66 @@
  */
 package com.snowplowanalytics.snowplow.sinks.kafka
 
-import cats.effect.Async
-import cats.effect.kernel.Resource
-import cats.implicits._
-import cats.Monad
-import com.snowplowanalytics.snowplow.sinks.{Sink, Sinkable}
-import fs2.kafka._
+import cats.effect.{Async, Resource, Sync}
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
+import org.apache.kafka.common.header.Header
+import org.apache.kafka.common.serialization.{ByteArraySerializer, StringSerializer}
 
-import scala.reflect._
+import com.snowplowanalytics.snowplow.sinks.{Sink, Sinkable}
+import com.snowplowanalytics.snowplow.azure.AzureAuthenticationCallbackHandler
 
 import java.util.UUID
-
-import com.snowplowanalytics.snowplow.azure.AzureAuthenticationCallbackHandler
+import java.nio.charset.StandardCharsets
+import scala.reflect.ClassTag
+import scala.jdk.CollectionConverters._
 
 object KafkaSink {
 
   def resource[F[_]: Async, T <: AzureAuthenticationCallbackHandler](
     config: KafkaSinkConfig,
     authHandlerClass: ClassTag[T]
-  ): Resource[F, Sink[F]] = {
-    val producerSettings =
-      ProducerSettings[F, String, Array[Byte]]
-        .withProperty("sasl.login.callback.handler.class", authHandlerClass.runtimeClass.getName)
-        .withBootstrapServers(config.bootstrapServers)
-        .withProperties(config.producerConf)
-
+  ): Resource[F, Sink[F]] =
     for {
-      producer <- KafkaProducer[F].resource(producerSettings)
-    } yield fromFs2Producer(config, producer)
+      producer <- makeProducer(config, authHandlerClass)
+    } yield impl(config, producer)
+
+  private def makeProducer[F[_]: Async, T <: AzureAuthenticationCallbackHandler](
+    config: KafkaSinkConfig,
+    authHandlerClass: ClassTag[T]
+  ): Resource[F, KafkaProducer[String, Array[Byte]]] = {
+    val producerSettings = Map(
+      "bootstrap.servers" -> config.bootstrapServers,
+      "sasl.login.callback.handler.class" -> authHandlerClass.runtimeClass.getName,
+      "key.serializer" -> classOf[StringSerializer].getName,
+      "value.serializer" -> classOf[ByteArraySerializer].getName
+    ) ++ config.producerConf
+    val make = Sync[F].delay {
+      new KafkaProducer[String, Array[Byte]]((producerSettings: Map[String, AnyRef]).asJava)
+    }
+    Resource.make(make)(p => Sync[F].blocking(p.close))
   }
 
-  private def fromFs2Producer[F[_]: Monad](config: KafkaSinkConfig, producer: KafkaProducer[F, String, Array[Byte]]): Sink[F] =
+  private def impl[F[_]: Sync](config: KafkaSinkConfig, producer: KafkaProducer[String, Array[Byte]]): Sink[F] =
     Sink { batch =>
-      val records = batch.copyToChunk.map(toProducerRecord(config, _))
-      producer.produce(records).flatten.void
+      Sync[F].interruptible {
+        val futures = batch.asIterable.map { e =>
+          val record = toProducerRecord(config, e)
+          producer.send(record)
```

Comment:
Do I get this right that in case of the buffer with pending records to get sent beig full, this call becomes blocking, which is why we also include it in the `Sync[F].interruptible { }` ?

Reply:
It becomes blocking under two circumstances:

1. The buffer is full, as you say.
2. The client needs to re-fetch topic metadata from the broker

Case 1 can be avoided by increasing the size of the buffer.  But case 2 is unavoidable.  So it must be run inside either `Sync[F].blocking` or `Sync[F].interruptible` (which are approximately the same thing).

Reply:
> 2. The client needs to re-fetch topic metadata from the broker

I remember a PR where you regularly manually fetch all the metadata in-parallel to avoid this blocking fetch. Do you now think that we should not do that ?

Reply:
I know the PR you are thinking of, but I abandoned it because I was wrong about it.  It is possible to pre-fetch the metadata when the app first starts, to avoid blocking on the first fetch.  But it is not possible to avoid blocking when the client decides to periodically re-fetch the metadata.